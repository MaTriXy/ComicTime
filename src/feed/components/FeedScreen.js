import React from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router";
import { fetchPublications, fetchNextPublicationsPage } from "../feed";
import NavigationBar from "../../baseComponents/navigationBar/NavigationBar";
import { translateComponent } from "../../i18n/i18n";
import ProgressBar from "../../baseComponents/progressBar/ProgressBar";
import { Scrollbars } from "react-custom-scrollbars";
import PublicationsFeed from "./PublicationsFeed";
import { publicationDetail } from "../../routes";

const style = {
  grid: {
    marginTop: 62
  }
};

const nextPageScrollBoundary = 0.75;

class FeedScreen extends React.Component {
  constructor(props) {
    super(props);
    this.onScrollFrame = this.onScrollFrame.bind(this);
    this.goToPublicationDetailScreen = this.goToPublicationDetailScreen.bind(
      this
    );
  }

  componentWillMount() {
    this.props.onComponentMounted();
  }

  render() {
    const { t } = this.props;
    if (this.props.loading) {
      return <ProgressBar />;
    } else {
      return (
        <div style={{ height: "100%", width: "100%", position: "absolute" }}>
          <NavigationBar title={t("appName")} />
          <Scrollbars style={style.grid} onScrollFrame={this.onScrollFrame}>
            <PublicationsFeed
              publications={this.props.publications}
              onPublicationSelected={id => this.goToPublicationDetailScreen(id)}
            />
          </Scrollbars>
        </div>
      );
    }
  }

  onScrollFrame(values) {
    if (values.top > nextPageScrollBoundary && !this.props.loadingNextPage) {
      this.props.onNextPageReached();
    }
  }

  goToPublicationDetailScreen(id) {
    const url = publicationDetail(id);
    this.props.history.push(url);
  }
}

export function mapStateToProps(state) {
  return {
    publications: state.feed.publications,
    loading: state.feed.fetchingPage && state.feed.publications.length === 0,
    loadingNextPage:
      state.feed.fetchingPage && state.feed.publications.length > 0
  };
}

export function mapPropsToDispatch(dispatch) {
  return {
    onComponentMounted: () => {
      dispatch(fetchPublications(1));
    },
    onNextPageReached: () => {
      dispatch(fetchNextPublicationsPage());
    }
  };
}

export const PureFeedScreen = FeedScreen;
export default withRouter(
  connect(mapStateToProps, mapPropsToDispatch)(
    translateComponent(PureFeedScreen)
  )
);
