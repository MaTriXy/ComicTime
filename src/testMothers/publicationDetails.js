import { IssueSummary, Publication } from "../publicationDetail/model";
import { Issue, IssuePage } from "../issueDetail/model";
import { Some } from "monet";

export const anyPublicationId = "the-flash-2016";
export const anyIssueId = "Annual1";

export const anyPublication = new Publication(
  "the-flash-2016",
  "https://comicimages.com/the-flash-1",
  "The Flash",
  Some("Ongoing"),
  "The best comic series in the world",
  Some("DC"),
  Some("January 2016"),
  Some("Peter Gomez"),
  [
    new IssueSummary("1", "the-flash-2016", "The Flash #1", "January 2016"),
    new IssueSummary("2", "the-flash-2016", "The Flash #2", "February 2016")
  ]
);

export const anyIssueSummary = new IssueSummary(
  "1",
  "detective-comic",
  "Detective Comic #1",
  "January 2018"
);

export const anyIssue = new Issue(
  "1",
  "detective-comic",
  "Detective Comic #1",
  2,
  [
    new IssuePage(1, "https://comicpages.com/1"),
    new IssuePage(2, "https://comicpages.com/2")
  ]
);
