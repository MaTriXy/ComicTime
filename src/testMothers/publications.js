import { Page, PublicationSummary } from "../feed/model";

export const anyListOfPublications = [
  new PublicationSummary(
    "uuid1",
    "http://readcomicsonline.ru/comic/deadpool-2017",
    "deadpool-2017",
    "deadpool (2017)",
    ["deadpool (2017) #1"],
    "http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg",
    "20/1/2017",
    "Marvel"
  ),
  new PublicationSummary(
    "uuid2",
    "http://readcomicsonline.ru/comic/volcanosaurus-2017",
    "volcanosaurus-2017",
    "Volcanosaurus (2017)",
    ["Volcanosaurus (2017) #2"],
    "http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg",
    "04/5/2018",
    "American Mythology"
  )
];

export const anyFirstPageOfPublications = new Page(anyListOfPublications, 1, 1);
export const anySecondPageOfPublications = new Page(
  anyListOfPublications,
  2,
  2
);
