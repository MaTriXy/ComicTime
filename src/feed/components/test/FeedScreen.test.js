jest.mock(
  "../../../baseComponents/navigationBar/NavigationBar",
  () => "NavigationBar"
);
jest.mock(
  "../../../baseComponents/progressBar/ProgressBar",
  () => "ProgressBar"
);
jest.mock("../PublicationsFeed", () => "PublicationsFeed");
jest.mock("../../apiClient/feedApiClient");
jest.mock("../../../i18n/i18n");

import React from "react";
import { PureFeedScreen } from "../FeedScreen";
import {
  shallowComponent,
  translator,
  shallowComponentAsJson
} from "../../../testUtils/enzyme";
import { anyListOfPublications } from "../../../testMothers/publications";

describe("FeedScreen", () => {
  it("renders a list of publications if it is not loading", () => {
    const component = shallowComponentAsJson(
      <PureFeedScreen
        t={translator}
        onComponentMounted={jest.fn()}
        loading={false}
        publications={anyListOfPublications}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("renders the progress bar if it is loading", () => {
    const component = shallowComponentAsJson(
      <PureFeedScreen
        t={translator}
        onComponentMounted={jest.fn()}
        loading={true}
      />
    );

    expect(component).toMatchSnapshot();
  });

  it("invokes the onComponentMounted props when it is mounted", () => {
    const onComponentMounted = jest.fn();

    shallowComponent(
      <PureFeedScreen
        t={translator}
        onComponentMounted={onComponentMounted}
        loading={true}
      />
    );

    expect(onComponentMounted).toBeCalled();
  });

  it("invokes the onNextPageReached props when the scroll is close to the end if not loading the next page", () => {
    const onNextPageReached = jest.fn();

    const component = shallowComponent(
      <PureFeedScreen
        t={translator}
        onComponentMounted={jest.fn()}
        onNextPageReached={onNextPageReached}
        loading={false}
        loadingNextPage={false}
      />
    );
    component.instance().onScrollFrame({ top: 0.76 });

    expect(onNextPageReached).toBeCalled();
  });

  it("does not invoke the onNextPageReached props when the scroll is close to the end if is loading the next page", () => {
    const onNextPageReached = jest.fn();

    const component = shallowComponent(
      <PureFeedScreen
        t={translator}
        onComponentMounted={jest.fn()}
        onNextPageReached={onNextPageReached}
        loading={false}
        loadingNextPage={true}
      />
    );
    component.instance().onScrollFrame({ top: 0.76 });

    expect(onNextPageReached).not.toBeCalled();
  });

  it("does not invokes the onNextPageReached props when the scroll is not close to the end even if not loading the next page", () => {
    const onNextPageReached = jest.fn();

    const component = shallowComponent(
      <PureFeedScreen
        t={translator}
        onComponentMounted={jest.fn()}
        onNextPageReached={onNextPageReached}
        loading={false}
        loadingNextPage={false}
      />
    );
    component.instance().onScrollFrame({ top: 0.75 });

    expect(onNextPageReached).not.toBeCalled();
  });
});
