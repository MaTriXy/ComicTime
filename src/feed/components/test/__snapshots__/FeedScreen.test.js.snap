// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FeedScreen renders a list of publications if it is not loading 1`] = `
<div
  style={
    Object {
      "height": "100%",
      "position": "absolute",
      "width": "100%",
    }
  }
>
  <NavigationBar
    title="appName"
  />
  <Scrollbars
    autoHeight={false}
    autoHeightMax={200}
    autoHeightMin={0}
    autoHide={false}
    autoHideDuration={200}
    autoHideTimeout={1000}
    hideTracksWhenNotNeeded={false}
    onScrollFrame={[Function]}
    renderThumbHorizontal={[Function]}
    renderThumbVertical={[Function]}
    renderTrackHorizontal={[Function]}
    renderTrackVertical={[Function]}
    renderView={[Function]}
    style={
      Object {
        "marginTop": 62,
      }
    }
    tagName="div"
    thumbMinSize={30}
    universal={false}
  >
    <PublicationsFeed
      publications={
        Array [
          PublicationSummary {
            "id": "uuid1",
            "image": "http://readcomicsonline.ru/uploads/manga/deadpool-2017/cover/cover_250x350.jpg",
            "lastIssuesNumbers": Array [
              "#1",
            ],
            "lastIssuesTitles": Array [
              "deadpool (2017) #1",
            ],
            "path": "deadpool-2017",
            "publisher": "Marvel",
            "releaseDate": "20/1/2017",
            "title": "deadpool (2017)",
            "url": "http://readcomicsonline.ru/comic/deadpool-2017",
          },
          PublicationSummary {
            "id": "uuid2",
            "image": "http://readcomicsonline.ru/uploads/manga/volcanosaurus-2017/cover/cover_250x350.jpg",
            "lastIssuesNumbers": Array [
              "#2",
            ],
            "lastIssuesTitles": Array [
              "Volcanosaurus (2017) #2",
            ],
            "path": "volcanosaurus-2017",
            "publisher": "American Mythology",
            "releaseDate": "04/5/2018",
            "title": "Volcanosaurus (2017)",
            "url": "http://readcomicsonline.ru/comic/volcanosaurus-2017",
          },
        ]
      }
    />
  </Scrollbars>
</div>
`;

exports[`FeedScreen renders the progress bar if it is loading 1`] = `<ProgressBar />`;
